### PR TITLE
Feature/slider

### PR DESCRIPTION
**Pull request by oleg331**

### CODE REVIEW

File: assets/js/slider.js [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+
+  const slider = function (settings) {
+
+    this.state = {
+      target: settings.target,
+      dotsContainer: settings.dotsContainer,
+      arrowLeft: settings.arrowLeft,
+      arrowRight: settings.arrowRight,
+      transition: settings.transition,
+      swipe: true,
```
Comment (kobweb) [inline assets/js/slider.js]: `swipe` and `autoHeigh` should be configurable too

File: assets/js/slider.js [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+    this.setDot();
+    if (this.state.autoHeight) {
+      this.state.target.style.height = this.allSlides[this.currentSlide].offsetHeight + "px";
+    }
+  }
+
+  slider.prototype.init = function () {
+    const self = this;
+
+    function on_resize(c, t) {
```
Comment (kobweb) [inline assets/js/slider.js]: camel case should be used

File: assets/js/slider.js [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+      onresize = function () {
```
Comment (kobweb) [inline assets/js/slider.js]: You forgot `const/let` ...

File: assets/js/slider.js [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+      return onresize;
+    }
+
+    window.addEventListener("resize", on_resize(() => {
+      self.updateSliderDimension();
+    }), false);
+
+    // Wrap slider-inner
+    const currentHTML = this.state.target.innerHTML;
+    this.state.target.innerHTML = '<div class="slider-inner">' + currentHTML + '</div>';
```
Comment (kobweb) [inline assets/js/slider.js]: Use es6 string templates

File: assets/js/slider.js [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+    this.allSlides = 0;
+    this.currentSlide = 0;
+    this.currentLeft = 0;
+    this.totalSlides = this.state.target.querySelectorAll('.slider-slide').length;
+
+    this.sliderInner = this.state.target.querySelector('.slider-inner');
+    this.loadedCnt = 0;
+
+    // Append clones
+    const cloneFirst = this.state.target.querySelectorAll('.slider-slide')[0].cloneNode(true);
```
Comment (kobweb) [inline assets/js/slider.js]: It would fail at runtime if there wouldn't be any `.slider-slide'` elements

File: assets/js/slider.js [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+    this.sliderInner.appendChild(cloneFirst);
+
+    const cloneLast = this.state.target.querySelectorAll('.slider-slide')[this.totalSlides - 1].cloneNode(true);
```
Comment (kobweb) [inline assets/js/slider.js]: The same as above

File: assets/js/slider.js [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+    }
+
+    function removeListenerMulti(elem, eventNames, eventFunc) {
+      eventNames.split(' ').forEach(function (e) {
+        return elem.removeEventListener(e, eventFunc, false);
+      });
+    }
+
+    if (this.state.swipe) {
+      addListenerMulti(this.sliderInner, 'mousedown touchstart', startSwipe);
```
Comment (kobweb) [inline assets/js/slider.js]: It's better to pass array of events instead of string.

File: assets/js/slider.js [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+    for (let i = 0; i < self.totalSlides + 2; i++) {
+      this.allSlides[i].style.width = 100 / (this.totalSlides + 2) + "%";
+      this.updateSliderDimension();
+    }
+
+    this.buildDots();
+    this.setDot();
+    this.initArrows();
+
+    function addListenerMulti(elem, eventNames, eventFunc) {
```
Comment (kobweb) [inline assets/js/slider.js]: Extract `addListenerMulti ` and `removeListenerMulti ` from the slider class. It's only a helper utils

File: assets/js/slider.js [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+      });
+    }
+
+    if (this.state.swipe) {
+      addListenerMulti(this.sliderInner, 'mousedown touchstart', startSwipe);
+    }
+
+    this.isAnimating = false;
+
+    function startSwipe(e) {
```
Comment (kobweb) [inline assets/js/slider.js:222]: Extract `startSwipe `, `swipeMove `, `swipeEnd ` as separated methods

File: assets/js/slider.js [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+      let touch = e;
+
+      if (e.type == 'touchmove') {
+        touch = e.changedTouches[0];
+      }
+      self.moveX = touch.pageX;
+      self.moveY = touch.pageY;
+
+      // for scrolling left and right
+      if (Math.abs(self.moveX - self.startX) < 40) return;
```
Comment (kobweb) [inline assets/js/slider.js:248]: It's a very bad practice to write `if` statements in one string.

File: assets/js/slider.js [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+      self.isAnimating = false;
+      removeClass(self.state.target, 'isAnimating');
+      removeListenerMulti(self.sliderInner, 'mousemove touchmove', swipeMove);
+      removeListenerMulti($('body'), 'mouseup touchend', swipeEnd);
+    }
+  }
+
+  slider.prototype.initArrows = function () {
+
+    this.state.arrowLeft.addEventListener('click', () => {
```
Comment (kobweb) [inline assets/js/slider.js]: Extract all event handler as separated methods

File: assets/js/slider.js [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+      if (!hasClass(this.state.target, 'isAnimating')) {
```
Comment (kobweb) [inline assets/js/slider.js]: Try to use a practice of early return
```
if (hasClass(this.state.target, 'isAnimating')) {
  return;
}
```

It makes your code simpler

File: assets/js/slider.js [hunk 13]
```diff
@@ -0,0 +1,283 @@
...
+          this.sliderInner.style.left = -this.currentSlide * this.slideW + 'px';
+        }
+        this.currentSlide++;
+        this.goToSlide();
+      }
+    }, false);
+  }
+
+  slider.prototype.setDot = function () {
+   let tardot = this.currentSlide - 1;
```
Comment (kobweb) [inline assets/js/slider.js]: WTF? Use normal readable variables names.

File: assets/js/slider.js [hunk 14]
```diff
@@ -0,0 +1,283 @@
...
+  slider.prototype.getCurrentLeft = function () {
+    this.currentLeft = parseInt(this.sliderInner.style.left.split('px')[0]);
+  }
+
+  slider.prototype.goToSlide = function () {
+
+    this.sliderInner.style.transition = 'left ' + this.state.transition.speed / 1000 + 's ' + this.state.transition.easing;
+    this.sliderInner.style.left = -this.currentSlide * this.slideW + 'px';
+
+    addClass(this.state.target, 'isAnimating');
```
Comment (kobweb) [inline assets/js/slider.js]: css class names shouldn't use camel case

File: assets/js/slider.js [hunk 15]
```diff
@@ -0,0 +1,283 @@
...
+
+    setTimeout(() => {
+      this.sliderInner.style.transition = '';
+      removeClass(this.state.target, 'isAnimating');
+    }, this.state.transition.speed);
+
+    this.setDot();
```
Comment (kobweb) [inline assets/js/slider.js]: I assume your dot would be set earlier than your slide due to `setTimeout`

File: assets/js/slider.js (hunk 1 above)
Comment (oleg331) [reply assets/js/slider.js]: Done

File: assets/js/slider.js (hunk 2 above)
Comment (oleg331) [reply assets/js/slider.js]: Done

File: assets/js/slider.js (hunk 3 above)
Comment (oleg331) [reply assets/js/slider.js]: Done

File: assets/js/slider.js (hunk 4 above)
Comment (oleg331) [reply assets/js/slider.js]: Done

File: assets/js/slider.js (hunk 5 above)
Comment (oleg331) [reply assets/js/slider.js]: Done

File: assets/js/slider.js (hunk 11 above)
Comment (oleg331) [reply assets/js/slider.js]: Done

File: assets/js/slider.js (hunk 12 above)
Comment (oleg331) [reply assets/js/slider.js]: Done

File: assets/js/slider.js (hunk 13 above)
Comment (oleg331) [reply assets/js/slider.js]: Done

File: assets/js/slider.js (hunk 14 above)
Comment (oleg331) [reply assets/js/slider.js]: Done

File: assets/js/slider.js (hunk 15 above)
Comment (oleg331) [reply assets/js/slider.js]: Done

**Pull request merged**